tests/v2/validation/observability/stackstate_test.go

```diff
@@ -0,0 +1,338 @@
+	defer subSession.Cleanup()
+
+	client, err := ss.client.WithSession(subSession)
+	require.NoError(ss.T(), err)
+
+	initialStackstateAgent, err := extencharts.GetChartStatus(client, ss.cluster.ID, charts.StackstateNamespace, charts.StackstateK8sAgent)
+	require.NoError(ss.T(), err)
+
+	if initialStackstateAgent.IsAlreadyInstalled {
+		ss.T().Skip("Stack state agent is already installed, skipping the tests.")
+	}
+
+	log.Info("Installing stack state agent on the provided cluster")
+
+	systemProject, err := projects.GetProjectByName(client, ss.cluster.ID, systemProject)
+	require.NoError(ss.T(), err)
+	require.NotNil(ss.T(), systemProject.ID)
+	systemProjectID := strings.Split(systemProject.ID, ":")[1]
+
+	err = charts.InstallStackstateAgentChart(ss.client, ss.stackstateAgentInstallOptions, ss.stackstateConfigs, systemProjectID)
+	require.NoError(ss.T(), err)
+	log.Info("Stack state chart installed successfully")
+
+	ss.T().Log("Verifying the deployments of stackstate agent chart to have expected number of available replicas")
+	err = extencharts.WatchAndWaitDeployments(client, ss.cluster.ID, charts.StackstateNamespace, meta.ListOptions{})
+	require.NoError(ss.T(), err)
+
+	ss.T().Log("Verifying the daemonsets of stackstate agent chart to have expected number of available replicas nodes")
+	err = extencharts.WatchAndWaitDaemonSets(client, ss.cluster.ID, charts.StackstateNamespace, meta.ListOptions{})
+	require.NoError(ss.T(), err)
+
+	clusterObject, _, _ := extensionscluster.GetProvisioningClusterByName(ss.client, ss.client.RancherConfig.ClusterName, fleet.Namespace)
+	if clusterObject != nil {
+		status := &provv1.ClusterStatus{}
+		err := steveV1.ConvertToK8sType(clusterObject.Status, status)
+		require.NoError(ss.T(), err)
+
+		podErrors := pods.StatusPods(client, status.ClusterName)
+		require.Empty(ss.T(), podErrors)
+	}
+}
+
+func (ss *StackStateTestSuite) TestUpgradeStackstateAgentChart() {
+	subSession := ss.session.NewSession()
+	defer subSession.Cleanup()
+
+	client, err := ss.client.WithSession(subSession)
+	require.NoError(ss.T(), err)
+
+	versionsList, err := client.Catalog.GetListChartVersions(charts.StackstateK8sAgent, rancherPartnerCharts)
+	require.NoError(ss.T(), err)
+
+	if len(versionsList) < 2 {
+		ss.T().Skip("Skipping the upgrade case, only one version of stackstate agent chart is available")
+	}
+
+	versionLatest := versionsList[0]
+	ss.T().Log(versionLatest)
+	versionBeforeLatest := versionsList[1]
+	ss.T().Log(versionBeforeLatest)
+	ss.stackstateAgentInstallOptions.Version = versionBeforeLatest
+
+	ss.T().Log("Checking if the stackstate agent chart is installed with one of the previous versions")
+	initialStackstateAgent, err := extencharts.GetChartStatus(client, ss.cluster.ID, charts.StackstateNamespace, charts.StackstateK8sAgent)
+	require.NoError(ss.T(), err)
+
+	if initialStackstateAgent.IsAlreadyInstalled || initialStackstateAgent.ChartDetails.Spec.Chart.Metadata.Version == versionLatest {
```

Comment:
is there a chance of nil pointer exception in the 2nd part of this statement?

Reply:
good catch it could lead to a nil pointer. removed the second check and the test will be skipped anyway if the first condition is true.